### ISSUE TITLE
Styling issues and missing pages when exporting to a PDF on Mozilla Firefox

### ISSUE THREAD

**Issue by jackcoble** [cryptee/web-client#108]
**Describe the bug**
When exporting a relatively large document to a PDF file (17 pages full of text and images) on Mozilla Firefox, there are some styling issues where content cuts off and overlaps onto the next page. There is also an issue towards the end of the PDF where it just shows blank pages instead of the desired content.

**To Reproduce**
1. Open up Mozilla Firefox
2. Create a document filled with several pages of text and images
3. Click on the document logo in the menu on the far right
4. Click on 'export/print', and then 'print/pdf'.
5. In the print preview, there may be the occasional content glitch where its split in half onto another page, making it difficult to read (please see the screenshots). However, the entire contents of the document are present in this print preview.

**Expected behavior**
When saving/exporting to PDF, content such as images should be placed onto a new page if they are too big, rather than splitting them across two pages.

Additionally, I would expect all of the document content to be exported in the PDF. When inspecting the exported PDF file, some of the content is missing completely towards the end.

**Screenshots**
![image](https://user-images.githubusercontent.com/26122014/109163029-c8923680-7770-11eb-9434-8d8a341de0b5.png)
Textual content occasionally being split onto a new page which makes it hard to read.

![image](https://user-images.githubusercontent.com/26122014/109162937-ae585880-7770-11eb-9416-3005e367d42a.png)
Images are being split/cut and spread over two pages. Instead I would expect if an image is too large for its current page, then it would be moved onto a new page.

![image](https://user-images.githubusercontent.com/26122014/109165128-1d36b100-7773-11eb-81ef-0476d3c773ce.png)
The print preview shows the entire document content. But when inspecting the exported PDF file, some content is missing completely towards the end of the document.

**System Information (please complete the following information):**
- Device: Lenovo Thinkpad X1 Carbon (Laptop non-touchscreen)
- OS: Linux Mint 20 - Cinnamon 
- Browser: Mozilla Firefox
- Browser Version: 85
- Browser Extensions: uBlock Origin, ClearURLs 
 - Do you have any DNS-based content/ad-blockers? No
 - Are you on a managed network? No

**Additional context**
I only seem to experience this issue using Mozilla Firefox. To make sure that it wasn't the document content or anything of that matter, I exported the document to a PDF file using Chromium. The exported document came out perfectly - there were none of the issues which I had experienced trying to export it using Firefox.

To make sure that my browser addons were not causing an issue between Cryptee and Firefox, I removed them completely and tried to export the document to a PDF file again. However this appeared to make no difference.

**Comment by jackcoble** [comment]
Just to make testing a bit easier, please find attached a Cryptee document containing titles, text and images of some puppies! 🐶
And then the exported PDF made in Firefox itself.

The password for the encrypted document is: **password123**

[test.ecd.zip](https://github.com/cryptee/web-client/files/6043991/test.ecd.zip)
[test_doc.pdf](https://github.com/cryptee/web-client/files/6044007/test_doc.pdf)

Some additional screenshots for more context...
![image](https://user-images.githubusercontent.com/26122014/109176427-b9b28080-777e-11eb-93f9-03e45695e723.png)
Image splitting,

![image](https://user-images.githubusercontent.com/26122014/109176476-c3d47f00-777e-11eb-90fe-d1068c3c64ec.png)
Hard to read text.

![image](https://user-images.githubusercontent.com/26122014/109176501-cc2cba00-777e-11eb-9cfb-8d3c8b5cee06.png)
Exported PDF content missing.

Hope this helps a bit more 😎

**Comment by johnozbay** [comment]
Hey there! Thanks a lot for filing this! ✌🏻

Sadly, [this is a known, 8+ years old Firefox issue](https://stackoverflow.com/questions/10355951/when-printing-firefox-cuts-off-crops-images-if-theyre-larger-then-a-single-pri), and not a Cryptee specific bug. 

In the past, Firefox used to use [PDF.js](https://github.com/mozilla/pdf.js) to render printable PDFs from the contents of the page (which is what we use in our pop-up previewer actually), and it used to work amazingly well. 

But [since v82, they've been slowly rolling out a new PDF/Print engine](https://support.mozilla.org/en-US/kb/how-print-web-pages-firefox#:~:text=Note:%20We%20are%20gradually%20rolling%20out%20this%20improved%20print%20menu%20starting%20in%20Firefox%20version%2082.), and [with this week's v86 update they released this new print/pdf engine for everyone](https://www.mozilla.org/en-US/firefox/86.0/releasenotes/#note-788575). 

PDF / Print rendering from web content is INCREDIBLY tricky, and this is why we intentionally don't even attempt to do it in Cryptee ourselves, and hand it off to the browser where possible. 

I wrote about this extensively here : https://github.com/cryptee/web-client/issues/32#issuecomment-555061881

TLDR; Most PDF generators on the web work by taking a screenshot of the HTML content then slicing the screenshot for each page. In developers' defense, this works quite well for 99% of people's web-pdf-printing use cases like printing QR code tickets or concerts etc, since these can all easily be printed as a screenshot cropped into multiple pages. Unencrypted services like Google Docs can handle these rendering operations on their servers – since the servers can see the contents of your documents – But with Cryptee, because everything is encrypted on your device, we can't use any server magic, and everything needs to happen in your browser. To add to the complexity, our frontend is entirely open source, so that anyone can audit our code, but PDF itself is a proprietary format by Adobe, and not open source compatible per-se, so we can't really use their proprietary code in our app.  

We can, in theory, render our own pages one by one, with the correct padding & spacing so nothing gets cropped, but this means we have to calculate the width / height / fonts' actual rendered pixel sizes on each browser and OS. (not to mention knowing paper stock sizes and printer ppi resolutions etc). As crazy as all this sounds like, we've  actually been working on this for a few months now. It's quite labor intensive [~~sheer insanity~~], but it's coming along nicely. Once this is complete, we'll add a print-accurate-mode / continuous-mode to our editor, and will support both editing styles / modes. You can track the progress for this here : #58

But I'm afraid until that's out, we can't really compute / render pages individually and generate a PDF file in memory, and until then, I'm afraid Firefox's PDF output may be unsatisfactory for some use cases 😔

Hoping all this makes sense, and I could summarize a topic that could normally take 3 hours to talk through hahah

I'll close this issue for now – but please join the conversation @ #58 and let us know what features you'd like to see! We'll start rolling out some of these features in the very near future! 

Thanks ✌🏻

**Comment by jackcoble** [comment]
Thanks for taking the time and explaining this to me. It is something I was totally unaware of beforehand! I'll be sure to take a look over the issues/feature request which you've linked. 😀

**Comment by johnozbay** [comment]
You're very welcome! Thank you for the incredibly detailed report, screenshots and helpful files! All deeply appreciated! 🙏🏻
